### ISSUE TITLE
WebGLState conflicts

### ISSUE THREAD

**Issue by Aldrien-** [mrdoob/three.js#7758]
I'd like to share a problem I encountered today about WebGLState.

There are at least two ways to enable/disable some states. For exemple, it's possible to disable blending with these two lines :

```javascript
renderer.state.setBlending( THREE.NoBlending );
```

and

```javascript
renderer.state.disable( renderer.context.BLEND );
```

A conflict can occurs because only the first one is based on an internal saved state (called `currentBlending`) . So, if you enable blending with the second way, the first one won't be able to disable it

I know OpenGL state is something to carefully change but three.js shouldn't add additional complexity to it.

**Comment by mrdoob** [comment]
How do you think it should be?

**Comment by WestLangley** [comment]
Also, can you please explain your use case? What is your need for manipulating the renderer state from the application level?

**Comment by Aldrien-** [comment]
### My use case

I am creating a SSAO effect on a scene which has shadow mapping (based on your implementation) and some HUD. Typically, I have different kinds of render passes: 

- depth map from light (for shadow mapping)
- depth map from scene camera (for ssao)
- ssao map from depth map (for ssao)
- a scene to the screen viewport
- a quad render in front of the screen which adds my blurred ssao effect with blending
- some quad render in front of screen viewport to make HUD

Note : I didn't use `THREE.EffectComposer` or something based on that because of antialiasing.

So, I have to manage several renderings, several kinds of rendering, several OpenGL states like blending or face culling for example.

**How I encountered what I described yesterday ?**  
I wanted to be sure of which OpenGL state is enabled or not without reading through each three.js codeline. So I decided to manually enable blending before my post process rendering, then disable it before my HUDs rendering and finally enable it at the end because it's its default state.

Note: you can solve the associated problem by a full understanding of what happens and a full management of your GL state (what I prefer and what I've done) or you can reset your GL state wìth `renderer.resetGLState()`. But there is still something not very logical with the API.

### An example

This is a basic scene:

[http://jsfiddle.net/Aldrien/690kywj2/](http://jsfiddle.net/Aldrien/690kywj2/)

![http://s14.postimg.org/4u4veh6u9/Untitled.png](http://s14.postimg.org/4u4veh6u9/Untitled.png)

This is the same scene with a simple post processing effect and a HUD:

[http://jsfiddle.net/Aldrien/pockqaff/](http://jsfiddle.net/Aldrien/pockqaff/)

![http://s21.postimg.org/5jjnvaijb/Untitled2.png](http://s21.postimg.org/5jjnvaijb/Untitled2.png)

This is what I wanted:

[http://jsfiddle.net/Aldrien/Ldyat63u/](http://jsfiddle.net/Aldrien/Ldyat63u/)

![http://s9.postimg.org/v8ejfu9fj/Untitled3.png](http://s9.postimg.org/v8ejfu9fj/Untitled3.png)

You're going to tell me that I don't need to add any blending management to make it work and you'll be right :-) But I couldn't know it and I didn't do anything illogical.

In fact, I've just enabled blending before my second render call (which will actually enable it by itself) and I disabled/enabled blending for the third call which should globally be ineffective. It should globally be ineffective because there is no blending stuff in this third scene.

Especially, the three.js render function will call :

```javascript
    // opaque pass (front-to-back order)

    state.setBlending( THREE.NoBlending );
    renderObjects( opaqueObjects, camera, lights, fog );
```

which is consistant with my disable call, and my enable call is consistant with to the default state : 

```javascript
    this.init = function () {

        gl.clearColor( 0, 0, 0, 1 );
        gl.clearDepth( 1 );
        gl.clearStencil( 0 );

        this.enable( gl.DEPTH_TEST );
        gl.depthFunc( gl.LEQUAL );

        gl.frontFace( gl.CCW );
        gl.cullFace( gl.BACK );
        this.enable( gl.CULL_FACE );

        this.enable( gl.BLEND );
        gl.blendEquation( gl.FUNC_ADD );
        gl.blendFunc( gl.SRC_ALPHA, gl.ONE_MINUS_SRC_ALPHA );

    };
```

**What is the problem ?**  
Like I said before, there are two not equivalent ways to disable some GL state like blending at least. In details, at the end of the first "animate" call the GL blend state is enable by `renderer.state.enable( renderer.context.BLEND )` and the WebGLState `currentBlending` param is still equal to `THREE.NoBlending`. From the second "animate" call, the first render pass will try to disable the blending by `state.setBlending( THREE.NoBlending );` (see above) but that will fail because the blending is already disabled from the WebGLState object point of view. So, the opaque sphere is badly rendered.

I know it's tricky and it's why I take my time to code a typical example and explain how and why ;-)

### How it should be

From my point of view, the most important thing is to be consistent in the API. There are different solutions which are opinionated, it's why I think the decision is yours.

1. You could not check the current state of blending (`currentBlending`) and directly set GL stuff every time it's asked. I can't suggest you should do that because I don't know the cost of GL change between rendering passes and I think you've done what you've done for a reason.

2. You could always check the current value of each GL state before setting it in an unique way. It's a little bit painful...

3. Like 2. you could only check the current value of each GL state which can be set in a different way : blending, depth function, depth write, ...

4. You could warn users to use specific functions for some GL state like blending. In this case, I suggest you clean up some code like `_state.enable( _gl.BLEND );` in `THREE.WebGLShadowMap` to not make users confused.

Ty for reading =D

**Comment by WestLangley** [comment]
@Aldrien- Thank you for your detailed explanation.

As you explained in your first post, the renderer can be placed into an internally-inconsistent state if the user calls `renderer.state.disable( renderer.context.BLEND );`.

@mrdoob Maybe that method should be "private"?

**Comment by mrdoob** [comment]
Something that would solve the issue would be making `WebGLState.enable/disable` set `currentBlending` to `null`. But it's sneaky and could potentially cause more problems...

**Comment by mrdoob** [comment]
> You could warn users to use specific functions for some GL state like blending. In this case, I suggest you clean up some code like _state.enable( _gl.BLEND ); in THREE.WebGLShadowMap to not make users confused.

A PR would be most welcome!

**Comment by WestLangley** [comment]
Any PR that prevents the user from placing the renderer into an invalid state is fine with me. The renderer needs to maintain tight control over its own state.

**Comment by Aldrien-** [comment]
I'll do some PR ;)

**Comment by Aldrien-** [comment]
More generally, as long as `gl` is public it's possible to change any GL state and make `WebGLState` "invalid". Each private `currentSomething` variable can be out of sync with its associated GL value. Moreover, `WebGLState.enable/disable` has its own private checking with `capabilities` which add another layer of complexity and another possibility of invalidity through a direct `gl` access.

So, before doing something I would like to know if you want to only focus on `WebGLState` state and how it manages itself or if you want to fully sync `WebGLState` with true `GL` states ?

**Comment by WestLangley** [comment]
@Aldrien- I know you would like more direction, but, if you don't mind, you may have to file a PR and then let @mrdoob, and others, respond to it.

**Comment by Mugen87** [comment]
Hi @Aldrien- ,

as you can see i have similar requirements ([#7785](https://github.com/mrdoob/three.js/issues/7785)). It would be great if the PR would also provide an interface for secure manipulation of the stencil buffer. If you need support or a tester, i'll help you if you like.

**Comment by Mugen87** [comment]
Sorry, I couldn't resist :wink:. I enhanced the `THREE.WebGLState` so users have an API to control the stencil buffer and don't need to access the native `WebGLRenderingContext`. Please have a look at this [jsFiddle](http://jsfiddle.net/g29k91qL/1/), especially the render method, where the relevant invocations are done. If you think the API is okay like this, i transfer this stuff to a PR.

`THREE.WebGLState` has the following new methods ( check out the code in this [file](http://yume.human-interactive.org/test/three.js)):

`setStencilFunc`, `setStencilOp`,`setStencilTest` and `setStencilWrite`

the file `Three.js` has the following new constants:

```javascript

// stencil functions

THREE.NeverStencil = 0;
THREE.AlwaysStencil = 1;
THREE.LessStencil = 2;
THREE.LessEqualStencil = 3;
THREE.EqualStencil = 4;
THREE.GreaterEqualStencil = 5;
THREE.GreaterStencil = 6;
THREE.NotEqualStencil = 7;

// stencil operations

THREE.KeepStencil = 0;
THREE.ReplaceStencil = 1;
THREE.IncrStencil = 2;
THREE.DecrStencil = 3;
THREE.InvertStencil = 4;
THREE.IncrWrapStencil = 5;
THREE.DecrWrapStencil = 6;

```

**Comment by mrdoob** [comment]
I think I would avoid creating javascript constants and use gl constants directly instead.

**Comment by Mugen87** [comment]
But that would mean users need access to native `WebGLRenderingContext`. 

Instead of
```javascript
renderer.state.setStencilFunc( THREE.AlwaysStencil, 1, 0xff );
```
users need to code this
```javascript
renderer.state.setStencilFunc( gl.ALWAYS, 1, 0xff );
```
As @Aldrien- said, public access to `WebGLRenderingContext` or the `gl`object can be problematic. If you think three.js should not encapsulate the entire WebGL interface, the mentioned constants are of course redundant.

Besides, i've tried to implemented the stencil buffer API like other parts of three.js. Depth buffer, blending etc. have also custom constants.

```javascript
THREE.NeverDepth = 0;
THREE.AlwaysDepth = 1;
THREE.LessDepth = 2;
THREE.LessEqualDepth = 3;
THREE.EqualDepth = 4;
THREE.GreaterEqualDepth = 5;
THREE.GreaterDepth = 6;
THREE.NotEqualDepth = 7;
```

**Comment by Aldrien-** [comment]
I've done some research about GL states and I agree with the WebGLState idea: the fewer GL state changes the better.

So, I've decided to only focus on how to fix the blending issue with this idea in mind and I split the blending management in two parts:

The first part is to manage the blending activation using WebGLState.enable/disable the same way DEPTH_TEST, POLYGON_OFFSET_FILL and SCISSOR_TEST work. In other words, the blending activation is managed by capabilities and not by currentBlending.
The second part is to let the blending parameters (blendEquation, blendFunc) be managed like it was.
I've built the library and it fixes the use case shown in this issue and doesn't seem to break examples, especially those with blending.
